module/state_synchronization/getter.go

```diff
@@ -0,0 +1,313 @@
+
+	done := make(chan struct{})
+	var v interface{}
+	var deserializeErr error
+
+	// start deserialization goroutine
+	go func() {
+		defer close(done)
+		defer bcr.Close()
+
+		if v, deserializeErr = s.serializer.Deserialize(bcr); deserializeErr != nil {
+			logger.Debug().Err(deserializeErr).Msg("failed to deserialize execution data")
+		}
+	}()
+
+	retrieveErr := s.retrieveBlobs(ctx, bs, cids, logger)
+
+	<-done
+
+	if retrieveErr != nil && !errors.Is(retrieveErr, blobs.ErrClosedBlobChannel) {
+		return nil, 0, retrieveErr
+	}
+
+	if deserializeErr != nil {
+		return nil, 0, &MalformedDataError{deserializeErr}
+	}
+
+	// TODO: deserialization succeeds even if the blob channel reader has still has unconsumed data, meaning that a malicious actor
+	// could fill the blob tree with lots of unnecessary data by appending it at the end of the serialized data for each level.
+	// It's possible that we could detect this and fail deserialization using something like the following:
+	// https://github.com/onflow/flow-go/blob/bd5320719266b045ae2cac954f6a56e1e79560eb/engine/access/rest/handlers.go#L189-L193
+	// Eventually, we will need to implement validation logic on Verification nodes to slash this behavior.
+
+	return v, bcr.TotalBytesRead(), nil
+}
+
+func (s *executionDataGetterImpl) getExecutionDataRoot(
+	ctx context.Context,
+	blockID flow.Identifier,
+	rootID flow.Identifier,
+	blobGetter network.BlobGetter,
+) (*ExecutionDataRoot, error) {
+	blob, err := blobGetter.GetBlob(ctx, flow.IdToCid(rootID))
+	if err != nil {
+		return nil, fmt.Errorf("failed to get execution data root blob: %w", err)
+	}
+
+	v, err := s.serializer.Deserialize(bytes.NewBuffer(blob.RawData()))
+	if err != nil {
+		return nil, fmt.Errorf("failed to deserialize execution data root blob: %w", err)
+	}
+
+	edRoot, ok := v.(*ExecutionDataRoot)
+	if !ok {
+		return nil, fmt.Errorf("execution data root blob is not an ExecutionDataRoot")
+	}
+
+	if edRoot.BlockID != blockID {
+		return nil, &MismatchedBlockIDError{blockID, edRoot.BlockID}
+	}
+
+	return edRoot, nil
+}
+
+func (s *executionDataGetterImpl) GetChunkExecutionDatas(
+	ctx context.Context,
+	blockID flow.Identifier,
+	rootID flow.Identifier,
+) ([]*ChunkExecutionData, error) {
+	logger := s.logger.With().Str("root_id", rootID.String()).Logger()
+	logger.Debug().Msg("getting execution data")
+
+	s.metrics.ExecutionDataGetStarted()
+
+	success := false
+	blobGetter := s.blobService.GetSession(ctx)
+	totalSize := atomic.NewUint64(0)
+	start := time.Now()
+
+	defer func() {
+		s.metrics.ExecutionDataGetFinished(time.Since(start), success, int(totalSize.Load()))
+	}()
+
+	edRoot, err := s.getExecutionDataRoot(ctx, blockID, rootID, blobGetter)
+	if err != nil {
+		return nil, err
+	}
+
+	g, gCtx := errgroup.WithContext(ctx)
+
+	chunkExecutionDatas := make([]*ChunkExecutionData, len(edRoot.ChunkExecutionDataIDs))
+	for i, chunkID := range edRoot.ChunkExecutionDataIDs {
+		i := i
+		chunkID := chunkID
+
+		g.Go(func() error {
```

Comment:
Can we explain the download strategy here?

From what I've understood. If a result has multiple chunks, the CIDs between chunks are concurrently downloaded. 

And for a single CID from a chunk, it will download level by level, until reaching `maxBlobTreeDepth`. For each level, the CIDs within the same level are downloaded _sequentially_. And we won't download CIDs from the next level, until CIDs from the current level are all downloaded. 

In other words, we only concurrently download CIDs for chunks, but within a single chunk, CIDs are downloaded sequentially. Do we consider downloading them concurrently as well?

Reply:
The CIDs within the same level are downloaded in parallel. This happens under the hood by Bitswap. It is not possible do download different levels in parallel, because the date from the previous level tells us what we need to download in the next level.

Reply:
Let's say the rootCID maps to 3 CIDs, so the first level has 3 CID: C1, C2, C3.

C1 has 2 CIDs: CA and CB.
C2 has 2 CIDs: CM and CN.
C3 has 2 CIDs: CX and CY.

so level 2 has 6 CIDs: [CA, CB, CM, CN, CX, CY].

If the getter has downloaded for C1, and learned C1 has two children: CA and CB, can we parallelize downloading for CA and CB while downloading C2 and C3? Why downloading CA from level 2 is blocked by downloading C2 in level 1?

Reply:
Ah, that's not quite how it works. What we do is we concatenate all the binary data we get from C1, C2, and C3, and then deserialize the entire piece of data into either another list of CID's, or an Execution Data.

So, to say that "C1 has 2 CIDs: CA and CB" is not accurate, because each CID on its own doesn't represent a self-contained piece of data. You need to concatenate all the data from an entire level in order to deserialize something that makes sense.

Reply:
I think what you have described might be conceptually cleaner, but it is somewhat more complex to implement, because now we need to handle cases such as, what to do if the subtree rooted at C1 has a different depth than the one rooted at C2? We should not allow this to happen, because otherwise it allows the structure of an Execution Data tree to be non-deterministic.

Reply:
In practice, the tree for a single ChunkExecutionData will be no more than 2 levels deep, because this already allows for ~32GB of data in total